cpp/modmesh/serialization/SerializableItem.hpp

```diff
@@ -55,65 +56,265 @@ namespace detail
 std::string escape_string(std::string_view str_view);
 
-template <typename T>
-std::string to_json_string(const T & value)
+/// Trim leading and trailing whitespaces and control characters.
+std::string trim_string(const std::string & str);
+
+/// State of the JSON parser.
+enum class JsonState
+{
+    Start,
+    ObjectKey,
+    Column,
+    Comma,
+    ObjectValue,
+    End,
+}; /* end enum class JsonState */
+
+/// Type of JSON token.
+enum class JsonType
+{
+    Object,
+    Array,
+    String,
+    Number,
+    Boolean,
+    Null,
+    Unknown,
+}; /* end enum class JsonType */
+
+struct JsonNode; // Forward declaration
+typedef std::unique_ptr<JsonNode> JsonNodePtr;
+typedef std::unordered_map<std::string, JsonNodePtr> JsonMap;
+typedef std::vector<JsonNodePtr> JsonArray;
+
+struct JsonNode
 {
-    if constexpr (std::is_base_of_v<SerializableItem, T>)
+    using JsonValue = std::variant<JsonMap, JsonArray, std::string>;
+
+    JsonType type;
+    JsonValue value;
+
+    JsonNode(JsonType type, const std::string & expression)
+        : type(type)
     {
-        return value.to_json(); /* recursive here */
+        parse(expression);
     }
-    else if constexpr (std::is_convertible_v<T, std::string>)
+
+private:
+
+    void parse(const std::string & expression)
     {
-        return "\"" + escape_string(value) + "\"";
+        if (type == JsonType::Object)
+        {
+            value = std::move(parse_object(expression));
+        }
+        else if (type == JsonType::Array)
+        {
+            value = std::move(parse_array(expression));
+        }
+        else
+        {
+            value = expression;
+        }
     }
-    else if constexpr (std::is_same_v<T, bool>)
+
+    static JsonMap parse_object(const std::string & json);
+
+    static JsonArray parse_array(const std::string & json);
+
+}; /* end struct JsonNode */
+
+/// Helper class for JSON serialization and deserialization.
+class JsonHelper
+{
+public:
+    template <typename T>
+    static std::string to_json_string(const T & value)
     {
-        return value ? "true" : "false";
+        if constexpr (std::is_base_of_v<SerializableItem, T>)
+        {
+            return value.to_json(); /* recursive here */
+        }
+        else if constexpr (std::is_convertible_v<T, std::string>)
+        {
+            return "\"" + escape_string(value) + "\"";
+        }
+        else if constexpr (std::is_same_v<T, bool>)
+        {
+            return value ? "true" : "false";
+        }
+        else if constexpr (is_specialization_of<std::vector, T>::value)
+        {
+            std::ostringstream oss;
+            oss << "[";
+            const char * separator = "";
+            for (const auto & item : value)
+            {
+                oss << separator << to_json_string(item); /* recursive here */
+                separator = ",";
+            }
+            oss << "]";
+            return oss.str();
+        }
+        else
```

Comment:
I am not familiar with `constexpr if` enough.  Do you need the keyword here?

Reply:
`constexpr if` is doing the condition in compilation time, which will not have a runtime branch.

Reply:
My question is against `else`.  Does it need `constexpr`?

Reply:
therefore, I think it's necessary here.

Reply:
> My question is against else. Does it need constexpr?

No need.